### PR TITLE
[node]: correct the return type for `destroy` and `end`

### PR DESCRIPTION
**Pull request by danielroe**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:

### `.end()`
* https://nodejs.org/api/stream.html#writableendchunk-encoding-callback
* https://nodejs.org/api/net.html#socketenddata-encoding-callback
* https://nodejs.org/api/http2.html#responseenddata-encoding-callback
* https://nodejs.org/api/http.html#requestenddata-encoding-callback
* https://nodejs.org/api/http.html#responseenddata-encoding-callback

### `.destroy()`
* https://nodejs.org/api/stream.html#readabledestroyerror
* https://nodejs.org/api/stream.html#writabledestroyerror
* https://nodejs.org/api/net.html#socketdestroyerror
* https://nodejs.org/api/http.html#messagedestroyerror
* https://nodejs.org/api/http.html#outgoingmessagedestroyerror

### Types corrected and verified with code

- [x] [ebml](https://github.com/node-ebml/node-ebml/blob/master/src/ebml/decoder.js#L11)
- [x] [hexo-util](https://github.com/hexojs/hexo-util/blob/master/lib/cache_stream.js#L5-L22)
- [x] [list-stream](https://github.com/rvagg/list-stream/blob/master/list-stream.js#L74-L81) - https://github.com/rvagg/list-stream/pull/3
- [x] [minipass](https://github.com/isaacs/minipass/blob/master/index.js#L226-L245)
- [x] [minizlib](https://github.com/isaacs/minizlib/blob/master/index.js#L108-L114)
- [x] [node-7z](https://github.com/quentinrossetti/node-7z/search?q=destroy)
- [x] [readable-stream](https://github.com/nodejs/readable-stream/blob/main/lib/internal/streams/destroy.js#L3-L54)
- [x] [firebird](https://github.com/xdenser/node-firebird-libfbclient/blob/master/firebird.js#L226-L238) - https://github.com/xdenser/node-firebird-libfbclient/pull/125
- [x] [ssh2](https://github.com/mscdex/ssh2/blob/master/lib/Channel.js#L214-L217) - https://github.com/mscdex/ssh2/pull/1096

### Libraries that implement `end`/`destroy` differently from the Node spec

These libraries' types will need changing to reflect the reality of what they are doing.

- [x] [bittorrent-protocol](https://github.com/webtorrent/bittorrent-protocol/blob/master/index.js#L46) - https://github.com/webtorrent/bittorrent-protocol/pull/74
- [x] [charm](https://github.com/substack/node-charm/blob/master/index.js#L103-L110)
- [x] [jsforce](https://github.com/jsforce/jsforce/blob/master/lib/query.js#L660-L722) - note that the type amended here wasn't correct in the first place - as this function always returns a Promise
- [x] [node-expat](https://github.com/astro/node-expat/blob/master/lib/node-expat.js#L74-L91) - https://github.com/astro/node-expat/pull/214
- [x] [request](https://github.com/request/request/blob/master/request.js#L1497-L1510)
- [x] [simple-peer](https://github.com/feross/simple-peer/blob/master/index.js#L430-L432) - https://github.com/feross/simple-peer/pull/855
- [x] [simple-websocket](https://github.com/feross/simple-websocket/blob/master/index.js#L99-L101)- https://github.com/feross/simple-websocket/pull/65
- [x] [tar-stream](https://github.com/mafintosh/tar-stream/blob/v1.6.2/extract.js#L192-L199)

### CODE REVIEW

**Comment by peterblazejewicz** [comment]
@danielroe 
I understand your change and change is OK, but it is expected to update and modify related types, like mentioned 'charm', so `@types/charm`, because with that change it should stop this PR on CI check:
```
Interface 'CharmInstance' incorrectly extends interface 'WritableStream'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
      ```

**Comment by danielroe** [comment]
I'd welcome any guidance as to how to handle the ecosystem of packages depending on `@types/node`, particularly the ones that don't implement Node 10+ interfaces fully.

I love the bot, by the way.

**Comment by sandersn** [comment]
I broke CI tests this morning when I updated the `@definitelytyped` packages. It is fixed now, but you'll have to merge from master to get the tslint changes to make tests pass.

**Comment by peterblazejewicz** [comment]
@danielroe 
fix CI (I'll rerun so one could see proper errors) and gather facts.
For example, `@types/charm`, this one source an be remove, this is Substack's module, already in long hiatus. This one can be removed from DT sources, while still on NPM registry. This means one less problem on CI (while still to discuss problem in broader terms). ONce done, please, re-request review

**Comment by danielroe** [comment]
@peterblazejewicz Thanks for rerunning CI. Looks like we're green ✅

I'll leave removing the charm types until you've had a chance to review my suggested solution for incompatible packages (using any for end/destroy return type, to avoid breaking these packages, for example when passing a stream to another function that expects a node compliant stream).

**Comment by amcasey** [comment]
@danielroe Thanks for the PR!  Your explanation makes sense and your change appears match your intentions, but I still have two concerns:

1. You had to fix up several older versions of node (thanks for doing so!) because the types have been like this for quite a long time, apparently without causing problems.  Do you have a guess about why this hasn't come up and been fixed before?  I can imagine that it might be because dotting into a destroyed object is unusual, but I'm not familiar with these types.
2. These types are *very* widely used, so I'd really like to hear from at least one @types/node owner and, ideally, representatives from some of the other packages as well (though I'm willing to stipulate that the cascading changes are likely correct if the node changes are correct).

**Comment by amcasey** [comment]
@jkomyno @alvis @r3nya @btoueg @smac89 @Touffy @DeividasBakanas @eyqs @Hannes-Magnusson-CK @hoo29 @kjin @ajafff @islishude @mwiktorczyk @mohsen1 @n-e @galkin @parambirs @eps1lon @westy92 @SimonSchick @ThomasdenH @WilcoBakker @wwwy3y3 @samuela @kuehlein @bhongy @chyzwar @trivikr @yoursunny @qwelias @ExE-Boss @Ryan-Willpower @peterblazejewicz @addaleax @victorperin @ZYSzys @nodejs @LinusU @wafuwafu13 

This change looks huge and scary, but it would be great if one or two people could give their thoughts on the general idea of changing the return types of these functions.  Thanks!

**Comment by karak** [comment]
As for `firebird`, returning `this` doesn't look to match with its JavaScript implementation.

Your changes:
<https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57473/files#diff-6f4afa29a12eeedb6221a3363d1d81bb121f49eae7b23d201e501da79c0fe868>

`.destroy()` returns nothing: <https://github.com/xdenser/node-firebird-libfbclient/blob/master/firebird.js#L201>
`.end()` also does: <https://github.com/xdenser/node-firebird-libfbclient/blob/master/firebird.js#L226>

It may confuse users to merge this part unless changing the original module.

**Comment by danielroe** [comment]
@karak Thanks for the review. My mistake. I've updated the return types in this PR to `any` for now and also submitted a PR upstream to update the implementation (https://github.com/xdenser/node-firebird-libfbclient/pull/125).

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## charm ([<kbd>unpkg</kbd>](https://unpkg.com/browse/charm@latest/))
was missing the following properties:
1. Charm
1. extractCodes
## jsforce ([<kbd>unpkg</kbd>](https://unpkg.com/browse/jsforce@latest/))
was missing the following properties:
1. addListener
1. on
1. once
1. removeListener
1. off

<details>
<summary>as well as these 29 other properties...</summary>
<p>removeAllListeners, setMaxListeners, getMaxListeners, listeners, rawListeners, emit, listenerCount, prependListener, prependOnceListener, eventNames, VERSION, require, addListener, on, once, removeListener, off, removeAllListeners, setMaxListeners, getMaxListeners, listeners, rawListeners, emit, listenerCount, prependListener, prependOnceListener, eventNames, VERSION, require</p>
</details>

## minipass ([<kbd>unpkg</kbd>](https://unpkg.com/browse/minipass@latest/))
was missing the following properties:
1. isStream
1. finished
1. pipeline
1. Stream
1. Readable

<details>
<summary>as well as these 4 other properties...</summary>
<p>Writable, Duplex, Transform, PassThrough</p>
</details>

## node-7z ([<kbd>unpkg</kbd>](https://unpkg.com/browse/node-7z@latest/))
was missing the following properties:
1. listenFactory
1. mainFactory
1. main
1. listenFactory
1. mainFactory

<details>
<summary>as well as these 1 other properties...</summary>
<p>main</p>
</details>

## readable-stream ([<kbd>unpkg</kbd>](https://unpkg.com/browse/readable-stream@latest/))
was missing the following properties:
1. from
1. finished
1. pipeline
1. once
1. on

<details>
<summary>as well as these 17 other properties...</summary>
<p>listenerCount, errorMonitor, captureRejectionSymbol, captureRejections, defaultMaxListeners, EventEmitter, from, finished, pipeline, once, on, listenerCount, errorMonitor, captureRejectionSymbol, captureRejections, defaultMaxListeners, EventEmitter</p>
</details>

## request ([<kbd>unpkg</kbd>](https://unpkg.com/browse/request@latest/))
was missing the following properties:
1. Request
1. Request
## ssh2 ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ssh2@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'ssh2'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.

The most common way to resolve this error is to use 'export =' syntax.
To learn more about 'export =' syntax, see https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require.
1. The declaration doesn't match the JavaScript module 'ssh2'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.

The most common way to resolve this error is to use 'export =' syntax.
To learn more about 'export =' syntax, see https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 6ad3deed6bd47e38feeb6116dc5a0c5b17953c22
</p>

**Review by dolanmiu — APPROVED** [review]

File: types/node/test/http2.ts [hunk 1]
```diff
@@ -321,14 +321,14 @@ import { URL } from 'node:url';
...
         response.write(Buffer.from([]), 'utf8', (err: Error) => {});
-        response.end();
-        response.end(() => {});
-        response.end('');
-        response.end('', () => {});
-        response.end('', 'utf8');
-        response.end('', 'utf8', () => {});
-        response.end(Buffer.from([]));
-        response.end(Buffer.from([]), () => {});
+        response.end().end();
```
Comment (Nokel81) [inline types/node/test/http2.ts:324]: Wouldn't it be a cleaner test if all of these lines were chained instead of just appending a `.end()` call?

**Review by Nokel81 — COMMENTED** [review]
Just a minor test comment. But this looks good to me.

**Comment by danielroe** [comment]
Ready to merge

**Comment by amcasey** [comment]
@vkarpov15 It's worth noting that DefinitelyTyped packages don't follow semver - major and minor come from the underlying library and patch is just incremented on every change.  Abstractly, breaking changes are generally accepted if they correct a bug, though that always needs to be balanced against the community impact.  In this particular case, I believe the impact was expected to be small because dotting into the return value of `destroy()` or `end()` was expected to be uncommon.  Do you have an example of where this caused problems?

**Comment by sandersn** [comment]
I'm not sure, but I think the problem is code that extends the affected types, like Readable. Overrides of `destroy` and `end` need to update their return types. 16 dependent projects needed to change on DT, which is a smallish change at node's scale.

**Comment by karak** [comment]
Finally, we have three options.

1.  All of the types has strict definitions like `end(): this` that some implementations aren't subject to. It might cause runtime errors.
2. Each type has AS-IS definitions. In this case, functions accepting `Readable` refuses some derivative types.
3. Base type has relaxed definitions like `end(): this | void`. It might miss some errors of type usage.

I think the third is the best choice (or to fix all implementations).

**Comment by LorneCurrie** [comment]
Is there a issue with this PR?  When I run a build in Typescript, I and getting the following errors:

```
backend/node_modules/@types/ebml/index.d.ts:233:5 - error TS2416: Property 'end' in type 'Decoder' is not assignable to the same property in base type 'Transform'.
  Type '{ (cb?: (() => void) | undefined): void; (chunk: Buffer, cb?: (() => void) | undefined): void; (chunk: Buffer, encoding?: string | undefined, cb?: (() => void) | undefined): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

233     end(cb?: () => void): void;
        ~~~

backend/node_modules/@types/ebml/index.d.ts:234:5 - error TS2416: Property 'end' in type 'Decoder' is not assignable to the same property in base type 'Transform'.
  Type '{ (cb?: (() => void) | undefined): void; (chunk: Buffer, cb?: (() => void) | undefined): void; (chunk: Buffer, encoding?: string | undefined, cb?: (() => void) | undefined): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.

234     end(chunk: Buffer, cb?: () => void): void;
        ~~~

backend/node_modules/@types/ebml/index.d.ts:235:5 - error TS2416: Property 'end' in type 'Decoder' is not assignable to the same property in base type 'Transform'.
  Type '{ (cb?: (() => void) | undefined): void; (chunk: Buffer, cb?: (() => void) | undefined): void; (chunk: Buffer, encoding?: string | undefined, cb?: (() => void) | undefined): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.

235     end(chunk: Buffer, encoding?: string, cb?: () => void): void;
        ~~~

```

**Pull request merged**